### PR TITLE
C#: Fix various extraction errors

### PR DESCRIPTION
**Pull request by calumgrant**
Fix 3 classes of extraction error:

1. Extracting multi-dimensional arrays, e.g. `byte *[][]` tripped up the typemention logic.
2. Not treating `_` as a regular variable name in a `foreach`, resulting in an unextracted local variable.
3. `nameof(System)`, where `System` is a namespace resulted in an unknown identifier type.

The last one required DB changes, because a `NameOfExpr` should ideally have a `getAccess()`, which in turn means that `Namespace` should be `@accessible` and we needed a new type of expression, `@namespace_access_expr` to represent this access. This is not super important but may as well be done for completeness.

### CODE REVIEW

**Comment by hvitved** [comment]
The unit tests are failing, I will have a look once the PR is green.

File: csharp/ql/src/semmle/code/csharp/Namespace.qll [hunk 1]
```diff
@@ -18,7 +18,7 @@ class TypeContainer extends DotNet::NamedElement, Element, @type_container { }
  * }
  * ```
  */
-class Namespace extends DotNet::Namespace, TypeContainer, @namespace {
+class Namespace extends DotNet::Namespace, TypeContainer, Declaration, @namespace {
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/Namespace.qll]: Should we add `DotNet::Declaration` as a base class of `DotNet::Namespace` as well?

File: csharp/ql/test/library-tests/exprorstmtparent/Declaration.expected [hunk 2]
```diff
@@ -40,3 +40,4 @@
 | B.cs:34:9:34:10 | M1 |
 | C.cs:1:15:1:16 | C4 |
 | C.cs:3:9:3:10 | M2 |
+| file://:0:0:0:0 |  |
```
[the remark this answers is not in the document]
Comment (calumgrant) [reply csharp/ql/test/library-tests/exprorstmtparent/Declaration.expected]: That's the location of the global namespace, which is now a declaration.

[the remark this answers is not in the document]
Comment (hvitved) [reply csharp/ql/test/library-tests/exprorstmtparent/Declaration.expected]: 👍

File: csharp/ql/src/semmle/code/csharp/Namespace.qll (hunk 1 above)
Comment (hvitved) [reply csharp/ql/src/semmle/code/csharp/Namespace.qll]: Come to think of it, we should run a dist-compare to check that performance is unchanged. (Changing the type hierarchy has resulted in unexpected optimizer behaviour before.)

**Comment by calumgrant** [comment]
Dist-compare results are as follows:
```
Project              fc0ffedda7a@1c51c3468  fc0ffedda7a@84fa0fb6d
corefx               4011                   3750                   0.934929
OrchardCore          2062                   2026                   0.982541
mono                 17798                  17739                  0.996685
OneOf                1541                   1537                   0.997404
PowerShell           1579                   1580                   1.00063
ql                   581                    584                    1.00516
EntityFrameworkCore  3464                   3496                   1.00924
coreclr              913                    929                    1.01752
roslyn               4403                   4487                   1.01908
```

**Pull request merged**